taichi/backends/cuda/cupti_toolkit.cpp

```diff
@@ -1,35 +1,832 @@
 #include "taichi/backends/cuda/cupti_toolkit.h"
+#include "taichi/backends/cuda/cuda_context.h"
+
+// move from cupti_toolkit.h
+// avoid exposing these headers
+#if defined(TI_WITH_CUDA_TOOLKIT)
+#include <cupti_target.h>
+#include <cupti_result.h>
+#include <cupti_profiler_target.h>
+#include <nvperf_host.h>
+#include <nvperf_cuda_host.h>
+#include <nvperf_target.h>
+#endif
 
 TLANG_NAMESPACE_BEGIN
 
-void CuptiToolkit::set_enable() {
-  cupti_config_.enable = true;
+// Make sure these metrics can be captured in one pass (no kernal replay)
+// Metrics for calculating the kernel elapsed time are collected by default.
+enum class CuptiMetricsDefault : uint {
+  CUPTI_METRIC_KERNEL_ELAPSED_CLK_NUMS = 0,
+  CUPTI_METRIC_CORE_FREQUENCY_HZS = 1,
+  CUPTI_METRIC_DEFAULT_TOTAL = 2
+};
+
+const char *MetricListDeafult[]{
```

Comment:
```suggestion
constexpr char *MetricListDeafult[] = {
```

Reply:
Thanks! I'll move on to finish CUPTI things.